rasa_core/tracker_store.py

```diff
@@ -250,3 +256,137 @@ def retrieve(self, sender_id):
+                 dialect: Text = 'sqlite',
+                 host: Text = None,
+                 event_broker: Optional[EventChannel] = None,
+                 db: Text = 'rasa.db',
+                 username: Text = None,
+                 password: Text = None) -> None:
+        from sqlalchemy.orm import sessionmaker
+        from sqlalchemy.engine.url import URL
+        from sqlalchemy import create_engine
+
+        engine_url = URL(dialect, username, password, host, database=db)
+
+        logger.debug('Attempting to connect to database '
+                     'via "{}"'.format(engine_url.__to_string__()))
+
+        self.engine = create_engine(engine_url)
+        self.session = sessionmaker(bind=self.engine)()
+
+        logger.debug('Connection successful, ensuring conversations table...')
+
+        self._ensure_table()
+        super(SQLTrackerStore, self).__init__(domain, event_broker)
+
+        logger.debug('SQL tracker store successfully initialised')
+
+    def _ensure_table(self) -> None:
+        """Creates the conversations table in the database if not present"""
+
+        metadata = MetaData()
+
+        Table("conversations", metadata,
+              Column("id", Integer, primary_key=True),
+              Column("sender_id", String, nullable=False),
+              Column("type_name", String, nullable=False),
+              Column("timestamp", Float),
+              Column("intent_name", String),
+              Column("action_name", String),
+              Column("data", String))
+
+        metadata.create_all(self.engine)
+
+    def keys(self) -> List:
+        """Returns the keys of the items stored in the database"""
+        return self.SQLEvent.__table__.columns.keys()
+
+    def retrieve(self, sender_id: Text) -> DialogueStateTracker:
+        """Creates a tracker from all previously stored events"""
+
+        subquery = self.session.query(self.SQLEvent)
+        query = subquery.filter_by(sender_id=sender_id).all()
+        events = [json.loads(event.data) for event in query]
+
+        if self.domain and len(events) > 0:
+            logger.debug('Recreating tracker '
+                         'from {} stored events'.format(len(events)))
+
+            tracker = DialogueStateTracker.from_dict(sender_id, events,
+                                                     self.domain.slots)
+        else:
+            logger.warning("Can't retrieve tracker from SQL storage.  "
+                           "Returning `None` instead.")
+            tracker = None
+        return tracker
+
+    def save(self, tracker: DialogueStateTracker) -> None:
+        """Updates database with events from the current conversation"""
+
+        if self.event_broker:
+            self.stream_events(tracker)
+
+        events = self._event_buffer(tracker)  # only store recent events
+
+        for event in events:
+            try:
+                data = event.as_dict()
+            except AttributeError as e:
```

Comment:
when does this happen? I don't think we do this anywhere else

Reply:
Happened during testing, but it has been fixed since `test_dialogues` were updated, I'll remove it now